c/tskit/trees.c

```diff
@@ -8145,3 +8165,339 @@ tsk_treeseq_extend_edges(const tsk_treeseq_t *self, int max_iter,
     tsk_table_collection_free(&tables);
     return ret;
 }
+
+/* ======================================================== *
+ * Pair coalescence
+ * ======================================================== */
+
+typedef struct {
+    /* TODO: better to include the implm from haplotype_matching.c, rather than
+     * duplicating? */
+    double value;
+    tsk_size_t index;
+} tsk_node_argsort_t;
+
+static int
+cmp_node_argsort(const void *a, const void *b)
+{
+    const tsk_node_argsort_t *ia = (const tsk_node_argsort_t *) a;
+    const tsk_node_argsort_t *ib = (const tsk_node_argsort_t *) b;
+    int ret = (ia->value > ib->value) - (ia->value < ib->value);
+    if (ret == 0) {
+        ret = (ia->index > ib->index) - (ia->index < ib->index);
+    }
+    return ret;
+}
+
+static int
+get_time_windows_index_map(const tsk_treeseq_t *self, tsk_size_t *num_time_windows,
+    const double *time_windows, tsk_id_t *result)
+{
+    int ret = 0;
+    const tsk_table_collection_t *tables = self->tables;
+    const tsk_size_t num_nodes = tables->nodes.num_rows;
+    tsk_node_argsort_t *nodes_time = NULL;
+    tsk_size_t i, j;
+
+    if (*num_time_windows == 0) { /* keep nodes unpooled */
+        for (i = 0; i != num_nodes; i++) {
```

Comment:
I guess it's the same thing, but ``for (i = 0; i < num_nodes; i++)`` is more idiomatic.

Reply:
Ah - I see where you got this from. The tree_pos inner loops are a special case because they can go in either direction, so we use != stop.